### PR TITLE
chore: clarify pull request process

### PR DESCRIPTION
**Pull request by latin-panda**
# Description

- clarity of including QA as reviewers
- clarify the importance of documentation as part of the development process
- Emphasis when a work is considered complete (ie, docs and test coverage are part of the process to completion). 
- suggests self-code reviews as best practices 

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

**Comment by latin-panda** [comment]
@andrablaj @mrjones-plip, Can you give me your thoughts on this? These are not new practices; most have been going on for a while. The intention is to clarify points like (QA, best practices, etc.) 

I'm leaving it as `Draft` because somehow my container gets stuck downloading Hugo's modules, so I'll verify the display later
Screenshot 2024-08-23 at 12 47 32 PM

**Comment by latin-panda** [comment]
It finally worked; this is the video of the docs in case it's easier to visualize the format. 


https://github.com/user-attachments/assets/fc7ab448-e0ea-469d-a6d5-0b3394b894f0

File: content/en/contribute/code/workflow.md [hunk 1]
```diff
@@ -173,9 +173,11 @@ Issues in this column have been prioritised and are ready for development. The i
...
 
-Issues in this column are being actively worked on, which includes development, design, code reviews, and testing.
+Issues in this column are being actively worked on, which includes development, design, documentation, code reviews, and testing. A work is considered complete when the documentation and test coverage (unit tests, e2e tests or integration tests) are finalized and the reviewers have approved it.
 
-Any code should be in a branch in each of the repositories you update. The name of the branch should be in the form `<issue-number>-<readable-name>`, for example `1104-inclusive-export`. Follow the [Quality Assistance]({{< ref "contribute/medic/product-development-process/quality-assistance" >}}) process to take full ownership of what you are building.
+#### Involve quality assistance early
+
+Involve [Quality Assistance]({{< ref "contribute/medic/product-development-process/quality-assistance" >}}) from the start of the process to take full ownership of what you are building.
 
 Use the following template for QA feedback throughout the development.
```
Comment (andrablaj) [inline content/en/contribute/code/workflow.md:182]: This was not part of your code changes, but I suggest moving this template under `Testing`, as this section here is about Project statuses and not so much about code review feedback or testing.

File: content/en/contribute/code/workflow.md [hunk 2]
```diff
@@ -251,21 +253,38 @@ The ticket needs further development.
...
+Create a Draft Pull Request to facilitate discussion and collaboration with quality assistance engineers and developers.
+
+Once you are confident that the change is complete and ready to be merged:
+
+1. Change the Pull Request from `Draft` to `Ready for review`.
+2. The Pull Request title will be the commit message, it is important to follow the [commit message format]({{< ref "#commit-message-format" >}}) to name the Pull Request title properly. 
+3. Add a Pull Request description:
 
-Once you're confident that the change is complete and ready to be merged:
+   - Add a description of changes, decisions, backstory, and any extra information to facilitate review process and reduce follow-ups.
```
Comment (andrablaj) [inline content/en/contribute/code/workflow.md:272]: This is overlapping a bit with the `Updating The Issue With What You Actually Did` section. I think we can merge them. I will write a suggestion in the final PR review comment.

File: content/en/contribute/code/workflow.md [hunk 3]
```diff
@@ -251,21 +253,38 @@ The ticket needs further development.
...
-1. Wait for the builds to succeed and ensure there are no conflicts with the the main branch so the PR can be merged.
-1. Pick one Reviewer for the PR and work with them until the code passes review. In some special cases more than one Reviewer may be necessary, but be specific about additional Reviewers and ensure you really need each of their additional reviews for a good reason. Remember, anyone can collaborate on PRs even if they aren't an official Reviewer. If you add a QA Engineer as a Reviewer, briefly comment in the ticket about what kind of testing review you expect from that engineer.
+4. Do a self-code review before asking for a review, this a good practice, almost always you will find things to fix. It saves a lot of time from you and the reviewers.
+5. Always assign a QA engineer and a developer as reviewers, they will give valuable feedback on their domain of expertise. It is okay to include one additional reviewer that has more experience in particular subject.
```
Comment (andrablaj) [inline content/en/contribute/code/workflow.md:277]: I want to challenge the "Always assign a QA engineer and a developer as reviewers". Always adding a QA reviewer shouldn't be necessary, and it puts a lot of pressure on our QA engineers to review _every_ PR. In an ideal setup, we should trust developers to share meaningful feedback about the testing done in a PR. I think the initial version of this bullet point is more accurate about how this process would happen: QA should be involved when the person writing the code needs QA expertise for some testing strategy. This might not always be the case, and that's fine too.

**Review by andrablaj — CHANGES_REQUESTED** [review]
Thanks for starting the discussion reviewing the workflow, I think it needs some attention and clarifications.

I added a few comments and I hope that I didn't bring too much confusion. To wrap-up my notes:

- The section testing template should live under the `Testing` section rather than under the `In Progress` one. 
- The `branch` and `pull request` sections under `In Progress` overlap significantly with other sections above. 
- I challenge the "Always add a QA Engineer as a reviewer" requirement. I explained the `why` in my comment. 
- The page is generally confusing (not because of this PR, to be clear 😅 ). We might need to review the sections, but that can be a separate PR. I see something like:
  - Issues
  - Project States
  - Writing
  - Branches
  - Commits
  - Testing
  - Pull requests
  - Code reviews

What do you think about this re-org of sections @latin-panda @mrjones-plip?

File: content/en/contribute/code/workflow.md (hunk 3 above)
Comment (latin-panda) [reply content/en/contribute/code/workflow.md:277]: I'm okay with that. I would like to add more clarity about this piece: 

> QA should be involved when the person writing the code needs QA expertise for some testing strategy 

When do developers add a QA as a reviewer, in your opinion? Asking for more specific scenarios
They can automate and test their work, which might make them think they don't see QA. In my experience. Quite often, they provide valuable feedback specially when it comes to their focus area (e2e / sql models, etc)

File: content/en/contribute/code/workflow.md (hunk 1 above)
Comment (latin-panda) [reply content/en/contribute/code/workflow.md:182]: Agree, this testing section was so confusing

File: content/en/contribute/code/workflow.md (hunk 3 above)
Comment (latin-panda) [reply content/en/contribute/code/workflow.md:277]: For example:

```
Consider including a QA engineer as a reviewer when you consider you need support for major changes or new features are being introduced to the codebase, security-related changes are made, substantial user experience improvements, etc.
```

File: content/en/contribute/code/workflow.md [hunk 4]
```diff
@@ -251,21 +253,38 @@ The ticket needs further development.
...
-1. Write a useful commit message in the PR using the [commit message format]({{< ref "#commit-message-format" >}}).
-2. Click the button to "Squash and Merge" the PR.
-3. If a backport is required cherry-pick the merged commit back to the release branches it's required in.
-4. Ensure the issue is added to the appropriate release milestone, which is the earliest semver version the change will be released in. This ensures it will be included in the release notes.
-5. Once all PRs have been merged, close the issue. This will automatically move it to "Done".
+1. If a backport is required cherry-pick the merged commit back to the release branches it's required in.
+2. Ensure the issue is added to the appropriate release milestone, which is the earliest semver version the change will be released in. This ensures it will be included in the release notes.
+3. Once all PRs have been merged, close the issue. This will automatically move it to "Done".
 
 ### Done
```
[the remark this answers is not in the document]
Comment (latin-panda) [reply content/en/contribute/code/workflow.md:289]: I've added these 3 bullet points to `Done` section

File: content/en/contribute/code/workflow.md (hunk 1 above)
Comment (latin-panda) [reply content/en/contribute/code/workflow.md:182]: Still weird, but better. When we restructure the page, then we'll improve it more

File: content/en/contribute/code/workflow.md (hunk 2 above)
Comment (latin-panda) [reply content/en/contribute/code/workflow.md:272]: Similar, but with a different audience, the PR usually contains more details for the reviewer, e.g the thinking process.
But I didn't explain it well here.

**Comment by latin-panda** [comment]
I agree this page needs to be restructured. In a separate PR next week?

File: content/en/contribute/code/workflow.md (hunk 3 above)
Comment (andrablaj) [reply content/en/contribute/code/workflow.md]: I agree with your sentence above. QA expertise can be constructive in the following scenarios: 

- The change includes user interface changes, and a QA engineer can advise whether that needs to be considered in the current e2e suite. 
- A developer has tested things locally and needs a second person to confirm that the expected behavior occurs on another machine. In this case, the developer should specifically request: "Can you please follow these steps and documentation and confirm that it is working as expected?"
- Any major changes/new features that require significant testing. The developer is expected to have written tests, but the QA Engineer can advise on edge cases or other scenarios to consider for testing.

**Comment by mrjones-plip** [comment]
I think @andrablaj is doing a great job here! pulling myself off reviewers list, but lemme know if you want me back for any reason!

File: content/en/contribute/code/workflow.md [hunk 5]
```diff
@@ -251,22 +180,114 @@ The ticket needs further development.
...
+6. Follow up on your PR, to keep momentum, the review should happen in 24h business days. If you haven't received any feedback from the reviewers after that time, check if they are available, otherwise it's okay to request a review from another person.
+7. Once the PR has been approved by a QA engineer and a developer, wait for the GitHub Actions to succeed and ensure there are no conflicts with the main branch.
+8. Double-check the [commit message format]({{< ref "#commit-message-format" >}}) is correct. Make sure to recognize collaboration in the commit description: `Co-authored-by: <gitnub user>`.
+9. Finally merge your work by selecting `Squash and merge`. This will compress all the commits into one, keeping the repository's commit history clean.
+   - If a backport is required cherry-pick the merged commit back to the release branches it's required in.
 
 ### Done
 
-Issues in this column are complete, all code has been merged into the main branch and/or release branches, and are ready for release.
+An issue is considered complete when:
```
[the remark this answers is not in the document]
Comment (latin-panda) [reply content/en/contribute/code/workflow.md:287]: Good catch!

File: content/en/contribute/code/workflow.md [hunk 6]
```diff
@@ -251,22 +180,114 @@ The ticket needs further development.
...
+ - The documentation is finalized
```
[the remark this answers is not in the document]
Comment (latin-panda) [reply content/en/contribute/code/workflow.md:288]: Good point, I'm going to add more clarification

> Does this mean that the doc PR needs to be fixed before the issue is considered Done?

IMO, the documentation (cht-docs, readme files, or comments) has to be ready before considering the work is "done", because: 
- The team is still on that topic -- writing documentation
- We are moving to community-first, and the documentation is our way to communicate how things work

File: content/en/contribute/code/workflow.md [hunk 7]
```diff
@@ -251,22 +180,114 @@ The ticket needs further development.
...
+   
+6. Follow up on your PR, to keep momentum, the review should happen in 24h business days. If you haven't received any feedback from the reviewers after that time, check if they are available, otherwise it's okay to request a review from another person.
+7. Once the PR has been approved by a QA engineer and a developer, wait for the GitHub Actions to succeed and ensure there are no conflicts with the main branch.
+8. Double-check the [commit message format]({{< ref "#commit-message-format" >}}) is correct. Make sure to recognize collaboration in the commit description: `Co-authored-by: <gitnub user>`.
+9. Finally merge your work by selecting `Squash and merge`. This will compress all the commits into one, keeping the repository's commit history clean.
+   - If a backport is required cherry-pick the merged commit back to the release branches it's required in.
 
 ### Done
 
-Issues in this column are complete, all code has been merged into the main branch and/or release branches, and are ready for release.
```
[the remark this answers is not in the document]
Comment (latin-panda) [reply content/en/contribute/code/workflow.md:272]: Do you mean a note like this at the end of the `Done` section?
Screenshot 2024-08-30 at 4 22 59 PM

File: content/en/contribute/code/workflow.md (hunk 3 above)
Comment (latin-panda) [reply content/en/contribute/code/workflow.md]: Thanks! I've added some of these in that paragraph

**Comment by latin-panda** [comment]
@andrablaj Thanks for all the thoughtful suggestions! 🤩 It's looking a lot better now

https://github.com/user-attachments/assets/430172aa-e3c8-498c-bfc5-1f9a6084dc4e

File: content/en/contribute/code/workflow.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (andrablaj) [reply content/en/contribute/code/workflow.md:272]: I was thinking of just keeping the paragraph as it is, without necessarily a `Note` formatting.

File: content/en/contribute/code/workflow.md [hunk 8]
```diff
@@ -251,22 +180,121 @@ The ticket needs further development.
...
+   - If a backport is required, cherry-pick the merged commit back to the release branches it is required in.
 
 ### Done
 
+An issue is considered complete when:
+ - The documentation is finalized.
+   - It includes merged updates in CHT Docs, updated README files, and necessary code comments.
+   - This is specially important as we care community-first and where they can find the latest details about CHT. 
+ - Unit tests, e2e tests and/or integration tests are written.
+ - The static checks like linting and Sonar should pass successfully as part of the quality process, any issues should be fixed.
```
Comment (andrablaj) [inline content/en/contribute/code/workflow.md:292]:
```suggestion
 - The static checks like linting and Sonar should pass successfully as part of the quality process, and any issues should be fixed. 
```

**Review by andrablaj — CHANGES_REQUESTED** [review]
Thanks for the updates! I added minor suggestions.

For some reason, the diff with the previous version got messed up and I am unsure what text was in the initial version versus now. I have some more comments about this page as a whole, but we can open a separate PR to re-shuffle some sections and make the reading experience more logical.

File: content/en/contribute/code/workflow.md [hunk 9]
```diff
@@ -251,22 +180,121 @@ The ticket needs further development.
...
+8. Double-check the [commit message format]({{< ref "#commit-message-format" >}}) is correct. Make sure to recognize collaboration in the commit description: `Co-authored-by: <gitnub user>`.
+9. Merge your work by selecting `Squash and merge`. This will compress all the commits into one, keeping the repository's commit history clean.
+   - If a backport is required, cherry-pick the merged commit back to the release branches it is required in.
 
 ### Done
 
+An issue is considered complete when:
+ - The documentation is finalized.
+   - It includes merged updates in CHT Docs, updated README files, and necessary code comments.
+   - This is specially important as we care community-first and where they can find the latest details about CHT.
```
[the remark this answers is not in the document]
Comment (andrablaj) [reply content/en/contribute/code/workflow.md:290]: I removed the `we` mention, per the style guidelines.

File: content/en/contribute/code/workflow.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (latin-panda) [reply content/en/contribute/code/workflow.md:272]: Ah got it! No worries, easy change

File: content/en/contribute/code/workflow.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (latin-panda) [reply content/en/contribute/code/workflow.md:290]: I always confuse specially with especially 🤦🏽‍♀️
Thanks!

**Comment by latin-panda** [comment]
@andrablaj I submitted the latest feedback, please have a look! Thanks!

**Review by andrablaj — APPROVED** [review]
Thanks @latin-panda for making this page clearer.

I created https://github.com/medic/cht-docs/issues/1517 to continue the conversation about reorganizing the sections on this page.
